src/main/java/net/fabricmc/fabric/mixin/client/rendercache/MixinBlockModelRenderer.java

```diff
@@ -0,0 +1,108 @@
+import net.minecraft.client.render.BufferBuilder;
+import net.minecraft.client.render.block.BlockModelRenderer;
+import net.minecraft.client.render.model.BakedModel;
+import net.minecraft.client.render.model.BakedQuad;
+import net.minecraft.util.crash.CrashException;
+import net.minecraft.util.crash.CrashReport;
+import net.minecraft.util.crash.CrashReportElement;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.ExtendedBlockView;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Random;
+
+@Mixin(BlockModelRenderer.class)
+public class MixinBlockModelRenderer {
+	// TODO NORELEASE: should this be split into two hooks to arrow overriding AO check?
+	@Inject(at = @At("HEAD"), method = "tesselate", cancellable = true)
+	public void tesselateSmoothRedir(ExtendedBlockView view, BakedModel model, BlockState state, BlockPos pos, BufferBuilder builder, boolean bool, Random random, long seed, CallbackInfoReturnable<Boolean> info) {
+		if (model instanceof DynamicBakedModel) {
+			boolean useAO = MinecraftClient.isAmbientOcclusionEnabled() && state.getLuminance() == 0 && model.useAmbientOcclusion();
+
+			try {
+				boolean result;
+				if (useAO) {
+					result = fabric_tesselateSmoothDynamic((RenderCacheView) view, (DynamicBakedModel) model, state, pos, builder, bool, random, seed);
+				} else {
+					result = fabric_tesselateFlatDynamic((RenderCacheView) view, (DynamicBakedModel) model, state, pos, builder, bool, random, seed);
+				}
+				info.setReturnValue(result);
+				info.cancel();
+			} catch (Throwable t) {
+				CrashReport report = CrashReport.create(t, "Tesselating dynamic block model");
+				CrashReportElement reportElementModel = report.addElement("Dynamic block being tesselated");
+				CrashReportElement.addBlockInfo(reportElementModel, pos, state);
+				reportElementModel.add("Using AO", useAO);
+				throw new CrashException(report);
+			}
+		}
+	}
+
+	public boolean fabric_tesselateSmoothDynamic(RenderCacheView view, DynamicBakedModel model, BlockState state, BlockPos pos, BufferBuilder builder, boolean allSides, Random rand, long seed) {
+		// TODO NORELEASE: AmbientOcclusionCalculator is non-public... ugh...
+		return fabric_tesselateFlatDynamic(view, model, state, pos, builder, allSides, rand, seed);
+	}
+
+	public boolean fabric_tesselateFlatDynamic(RenderCacheView view, DynamicBakedModel model, BlockState state, BlockPos pos, BufferBuilder builder, boolean allSides, Random rand, long seed) {
+		ExtendedBlockView eView = (ExtendedBlockView) view;
+		List<BakedQuad> quads = new ArrayList<>();
+		BitSet bitSet = new BitSet(3);
+		boolean rendered = false;
```

Comment:
It baffles me why Mojang chose to instantiate a new BitSet here, of all places, just to hold a few bit flags.  I can only assume their profiling showed escape analysis to be reliable here.

Reply:
It's Mojang; don't assume, but verify :thinking: